File: 2021/Day 7/TS/part2.ts
import fs from 'fs'
import path from 'path'

const file: string = fs.readFileSync(path.resolve(__dirname, '../test-data.txt'), 'utf8')
const positions = file.split(',').map((str) => parseInt(str))

// Find average, then try both values near it
const target1 = Math.floor(positions.reduce((sum, val) => sum + val, 0) / positions.length)
const targets = [target1, target1 + 1]
let fuel = targets
    .map((target) =>
        // Get fuel consumption for each possible value
        positions.reduce((sum, pos) => sum + (Math.abs(pos - target) * (Math.abs(pos - target) + 1)) / 2, 0),
    )
    // Get lowest fuel consumption of the two
    .reduce((min, val) => (val < min ? val : min))

console.log('Fuel consumption:', fuel)

File: 2021/Day 7/C++/part2.cpp
// Headers
#include <algorithm>
#include <array>
#include <cmath>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <vector>

// Namespaces
using namespace std;  // Bad practice, I know

// Prototypes

// Main
int main() {
    // Get input
    ifstream file("../data.txt");
    vector<int> positions;
    int position;
    while (file >> position) {
        positions.push_back(position);
        file.ignore();
    }

    // Calculate fuel consumption
    double average = 0;
    for (double pos : positions) average += pos / positions.size();
    const int targets[] = {int(floor(average)), int(floor(average)) + 1};
    int fuel = numeric_limits<int>::max();
    for (int target : targets) {
        int newFuel = 0;
        for (int pos : positions) newFuel += abs(pos - target) * (abs(pos - target) + 1) / 2;
        if (newFuel < fuel) fuel = newFuel;
    }

    // Output result
    cout << "Fuel consumption: " << fuel << endl;

    return 0;
}